### DISCUSSION TITLE
Should MachineContext.useSelector detect self-transitioned events?

### DISCUSSION THREAD

**Discussion by kindkitchen** [statelyai/xstate#4648] [Q&A] [answered]
In my react application the component miss some machine's context updates that are happen during self-transitioned events.
I mean the callback that I pass into `.useSelector()` is not triggered.
So now I need some clarifications aka:
* is it normal behavior?
* may be I use self-transition in wrong way?
P.S.
I use `.useSelector()` from ContextProvider (@xstate/react).

**Comment by kindkitchen** [comment]
Sorry, I too early posted this post... I still little confusing though)))

But I realize that my problem has a little another source and I think I should play around little more. My own question above is little strange for my now but I can't delete it)))

**Reply by davidkpiano** [reply]
You can edit the original post if you'd like

**Reply by kindkitchen** [reply]
Thanks, hope this question is little useful)

**Comment by Andarist** [comment, marked as answer]
I'm not sure what kind of detection you are interested in here but it usually shouldn't matter in React. You can't easily "respond" to a change from within `useSelector` - React's model is about synchronization to data sources. In it, it's totally fine to "miss" some updates as long as you receive a fresh one. 

IMHO, this matches React's state update batching quite close (see the docs around it [here](https://react.dev/learn/queueing-a-series-of-state-updates)):
> React waits until all code in the event handlers has run before processing your state updates. This is why the re-render only happens after all these setNumber() calls.

**Reply by kindkitchen** [reply]
__Thank you a lot!__ I know that I must already read this part of react documentation and I even thought that I understand this moment... Simple but so important moment! You opened my eyes)
Thank you again, really helpful for me!
(By the way after analyzing this moment I solve my issue using `assign` helper for updating context (yes... I didn't use it before))) and `.useSelector` worked only on changing states that's why I mistakenly thought that issue is related to self-transitions)
